plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusingConnectionFactory.java

```diff
@@ -0,0 +1,205 @@
+        this.connections = connections.asMap();
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.remove(queryId);
+        if (connection != null) {
+            return connection;
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.remove(session.getQueryId());
+        if (connection != null) {
+            try {
+                connection.close();
+            }
+            catch (SQLException e) {
+                throw new TrinoException(JDBC_ERROR, e);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        for (Connection connection : connections.values()) {
+            connection.close();
+        }
+        connections.clear();
+        delegate.close();
+    }
+
+    private final class CachedConnection
+            extends ForwardingConnection
+    {
+        private final String queryId;
+        private final Connection delegate;
+        private boolean closed;
+        private boolean dirty;
+
+        private CachedConnection(String queryId, Connection delegate)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        protected Connection getDelegate()
+        {
+            checkState(!closed, "Connection is already closed");
+            return delegate;
+        }
+
+        @Override
+        public <T> T unwrap(Class<T> iface)
+                throws SQLException
+        {
+            dirty = true;
+            return super.unwrap(iface);
+        }
+
+        @Override
+        public void setReadOnly(boolean readOnly)
+                throws SQLException
+        {
+            dirty = true;
+            super.setReadOnly(readOnly);
+        }
+
+        @Override
+        public void setCatalog(String catalog)
+                throws SQLException
+        {
+            dirty = true;
+            super.setCatalog(catalog);
+        }
+
+        @Override
+        public void setSchema(String schema)
+                throws SQLException
+        {
+            dirty = true;
+            super.setSchema(schema);
+        }
+
+        @Override
+        public void setTransactionIsolation(int level)
```

Comment:
what about `setSavepoint`?

Reply:
maybe we should allow only certain method as "not-dirting". This can be easily achieved using a Proxy.
I guess only SELECT queries are not making a connection dirty.
`createStatement(String)`, `createStatement().executeQuery`, `getDatabaseMetaData` ?